tests/pass-dep/concurrency/freebsd-futex.rs

```diff
@@ -0,0 +1,250 @@
+                    123,
+                    timeout_size_arg,
+                    &mut timeout as *mut _ as _,
+                ),
+                -1
+            );
+            assert_eq!(io::Error::last_os_error().raw_os_error().unwrap(), libc::ETIMEDOUT);
+        }
+        assert!((200..1000).contains(&start.elapsed().as_millis()));
+    }
+
+    without_timespec();
+    with_timespec();
+}
+
+fn wait_absolute_timeout() {
+    let start = Instant::now();
+
+    // Get the current monotonic timestamp as timespec.
+    let mut timeout = unsafe {
+        let mut now: MaybeUninit<libc::timespec> = MaybeUninit::uninit();
+        assert_eq!(libc::clock_gettime(libc::CLOCK_MONOTONIC, now.as_mut_ptr()), 0);
+        now.assume_init()
+    };
+
+    // Add 200ms.
+    timeout.tv_nsec += 200_000_000;
+    if timeout.tv_nsec > 1_000_000_000 {
+        timeout.tv_nsec -= 1_000_000_000;
+        timeout.tv_sec += 1;
+    }
+
+    // Create umtx_timeout struct with that absolute timeout.
+    let umtx_timeout = libc::_umtx_time {
+        _timeout: timeout,
+        _flags: libc::UMTX_ABSTIME,
+        _clockid: libc::CLOCK_MONOTONIC as u32,
+    };
+    let umtx_timeout_ptr = &umtx_timeout as *const _;
+    let umtx_timeout_size = ptr::without_provenance_mut(mem::size_of_val(&umtx_timeout));
+
+    let futex: u32 = 123;
+
+    // Wait for 200ms from now, with nobody waking us up early.
+    unsafe {
+        assert_eq!(
+            libc::_umtx_op(
+                ptr::from_ref(&futex).cast_mut().cast(),
+                libc::UMTX_OP_WAIT_UINT_PRIVATE,
+                123,
+                umtx_timeout_size,
+                umtx_timeout_ptr as *mut _,
+            ),
+            -1
+        );
+        assert_eq!(io::Error::last_os_error().raw_os_error().unwrap(), libc::ETIMEDOUT);
+    }
+    assert!((200..1000).contains(&start.elapsed().as_millis()));
+}
+
+fn wait_wake() {
+    // we create 2 threads that wait on a futex with a 500ms timeout.
+    // The main thread wakes up 1 thread waiting on this futex and after 100ms
+    // checks that only 1 thread woke up and the other timed out.
+    static mut FUTEX: u32 = 0;
+
+    fn waiter() -> bool {
+        let mut timeout = libc::timespec { tv_sec: 0, tv_nsec: 500_000_000 };
+        let timeout_size_arg =
+            ptr::without_provenance_mut::<libc::c_void>(mem::size_of::<libc::timespec>());
+        unsafe {
+            libc::_umtx_op(
+                addr_of!(FUTEX) as *mut _,
+                libc::UMTX_OP_WAIT_UINT_PRIVATE,
+                0, // FUTEX is 0
+                timeout_size_arg,
+                &mut timeout as *mut _ as _,
+            );
+            io::Error::last_os_error().raw_os_error().unwrap() == libc::ETIMEDOUT
+        }
+    }
+
+    let t1 = thread::spawn(waiter);
+    let t2 = thread::spawn(waiter);
+
+    // Wake up 1 thread and make sure the other is still waiting
+    thread::sleep(Duration::from_millis(200));
```

Comment:
Again no `sleep` needed. You should `yield_now` here once, then the two threads will be scheduled and block on the futex. Then you wake up one. There shouldn't be any `sleep` in this test.